### PR TITLE
[Feature] Adds value clipping in ClipPPOLoss loss

### PR DESCRIPTION
**Pull request by albertbou92**
## Description

This PR add a value clipping option in ClipPPOLoss loss.

This PR is related to https://github.com/pytorch/rl/issues/1977.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/objectives/ppo.py [hunk 1]
```diff
@@ -664,6 +664,8 @@ class ClipPPOLoss(PPOLoss):
             ``"none"`` | ``"mean"`` | ``"sum"``. ``"none"``: no reduction will be applied,
             ``"mean"``: the sum of the output will be divided by the number of
             elements in the output, ``"sum"``: the output will be summed. Default: ``"mean"``.
+        clip_value_loss (bool, optional): if ``True``, the value loss will be clipped with respect to the
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:667]: I would have this
- not provided / False = no clipping
- True = same clipping as log ratio
- number = clip with that number 

That way you can reuse the same semantic for other losses

**Comment by albertbou92** [comment]
We can potentially add it to more losses (PPOLoss, KLPENPPOLoss, A2C, Reinforce) with the only difference that we would require the user to provide a float and not a bool since we can not default to the log ratio clipping value. 

But I am not sure if we should see if the feature is asked first for those other losses, since I have never seen in used in those.

**Comment by vmoens** [comment]
I don't thinks it's a problem to add it, we don't want to copy existing repos but enable people to experiment and swap configurations quickly.

File: torchrl/objectives/reinforce.py [hunk 2]
```diff
@@ -433,6 +456,20 @@ def loss_critic(self, tensordict: TensorDictBase) -> torch.Tensor:
                 f"TDLambdaEstimate and TDEstimate all return a 'value_target' entry that "
                 f"can be used for the value loss."
             )
+
+        if self.clip_value_loss:
+            self.clip_value_loss = self.clip_value_loss.to(state_value.device)
```
Comment (vmoens) [inline torchrl/objectives/reinforce.py:461]: ditto, let's avoid the casting of the attribute but cast the variable only

File: torchrl/objectives/ppo.py [hunk 3]
```diff
@@ -524,6 +547,20 @@ def loss_critic(self, tensordict: TensorDictBase) -> torch.Tensor:
             state_value,
             loss_function=self.loss_critic_type,
         )
+
+        if self.clip_value_loss:
+            self.clip_value_loss = self.clip_value_loss.to(state_value.device)
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:552]: ditto, let's avoid the casting of the attribute but cast the variable only

**Review by vmoens — COMMENTED** [review]
Thanks for this! Cool feature! 
A couple of comments along the way but otherwise LGTM

**Comment by albertbou92** [comment]
I added the clip_fraction for both the PPO loss and the Value loss

File: torchrl/objectives/a2c.py [hunk 4]
```diff
@@ -474,8 +523,10 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
             td_out.set("entropy", entropy.detach().mean())  # for logging
             td_out.set("loss_entropy", -self.entropy_coef * entropy)
         if self.critic_coef:
-            loss_critic = self.loss_critic(tensordict)
+            loss_critic, value_clip_fraction = self.loss_critic(tensordict)
             td_out.set("loss_critic", loss_critic)
+            if value_clip_fraction is not None:
+                td_out.set("value_clip_fraction", value_clip_fraction)
```
Comment (vmoens) [inline torchrl/objectives/a2c.py:529]: we should include that key in the out_keys

File: torchrl/objectives/ppo.py [hunk 5]
```diff
@@ -506,6 +527,16 @@ def loss_critic(self, tensordict: TensorDictBase) -> torch.Tensor:
                 f"can be used for the value loss."
             )
 
+        if self.clip_value:
+            try:
+                old_state_value = tensordict.get(self.tensor_keys.value).clone()
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:532]: why do we clone here?

File: torchrl/objectives/a2c.py [hunk 6]
```diff
@@ -445,7 +475,26 @@ def loss_critic(self, tensordict: TensorDictBase) -> torch.Tensor:
...
+            )
+            # Chose the most pessimistic value prediction between clipped and non-clipped
+            loss_value = torch.max(loss_value, loss_value_clipped)
+            clip_fraction = (
+                (state_value / old_state_value)
+                .clamp(1 - clip_value, 1 + clip_value)
+                .abs()
+                .detach()
+            )
+        return self.critic_coef * loss_value, clip_fraction
```
Comment (vmoens) [inline torchrl/objectives/a2c.py:497]: That seems to be a lot of duplicated code, can we move that to objectives.utils within a private function?

File: torchrl/objectives/a2c.py [hunk 7]
```diff
@@ -445,7 +475,26 @@ def loss_critic(self, tensordict: TensorDictBase) -> torch.Tensor:
...
+                state_value - old_state_value
+            ).clamp(-clip_value, clip_value)
+            loss_value_clipped = distance_loss(
+                target_return,
+                state_value_clipped,
+                loss_function=self.loss_critic_type,
+            )
+            # Chose the most pessimistic value prediction between clipped and non-clipped
+            loss_value = torch.max(loss_value, loss_value_clipped)
+            clip_fraction = (
```
Comment (vmoens) [inline torchrl/objectives/a2c.py:491]: I don't get what clip_fraction represents

One thing I like to look at is what proportion of data is being clipped

Also we should be careful about this since it doesn't come for free. The overhead introduced by logging this metric could impact performance...

File: torchrl/objectives/a2c.py [hunk 8]
```diff
@@ -445,7 +475,26 @@ def loss_critic(self, tensordict: TensorDictBase) -> torch.Tensor:
...
+                (state_value / old_state_value)
+                .clamp(1 - clip_value, 1 + clip_value)
+                .abs()
+                .detach()
```
Comment (vmoens) [inline torchrl/objectives/a2c.py:495]: Instead of detach() I think we should put the whole thing within a torch.no_grad()

File: torchrl/objectives/ppo.py (hunk 5 above)
Comment (albertbou92) [reply torchrl/objectives/ppo.py:532]: Because the `self.tensor_keys.value` prediction will be recomputed once we pass tensordict through the value network and we will lose this tensor.

File: torchrl/objectives/a2c.py (hunk 7 above)
Comment (albertbou92) [reply torchrl/objectives/a2c.py:491]: Yes, as I understand it, it is a measure of how much the old model that collected the data differs from the current model. If the fraction values stay close to 1.0 the model is changing slowly.  If the ratio consistently deviates from 1.0 and approaches or hits the clipping value, it indicates that the new model is substantially different from the old policy, indicating a faster rate of policy change.

File: torchrl/objectives/a2c.py (hunk 8 above)
Comment (albertbou92) [reply torchrl/objectives/a2c.py:495]: Makes sense. Otherwise the oprations up the detach are still tracked by autograd right?

**Review by vmoens — APPROVED** [review]
LGTM
